doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+
+This will run correctly, but tracing and recompilation will occur with every
+function execution:
+
+>>> circuit(jnp.array([0.1, 0.2]))
+Tracing occurring
+array(0.99500417)
+>>> circuit(jnp.array([0.1, 0.2, 0.3]))
+Tracing occurring
+array(0.99500417)
+
+To be explicitly warned about recompilation, you can use ahead-of-time
+(AOT) mode, by specifying types and shapes in the function signature
+directly:
+
+>>> @qjit
+... @qml.qnode(dev)
+... def circuit(x: jax.core.ShapedArray((3,), dtype=np.float64)):
+...     print("Tracing occurring")
+...     def loop_fn(i):
+...         qml.RX(x[i], wires=i)
+...     for_loop(0, x.shape[0], 1)(loop_fn)()
+...     return qml.expval(qml.PauliZ(0))
+Tracing occurring
+
+Note that compilation now happens on **function definition**. We can execute
+the compiled function as long as the arguments match the specified shapes and
+type:
+
+>>> circuit(jnp.array([0.1, 0.2, 0.3]))
+array(0.99500417)
+>>> circuit(jnp.array([1.4, 1.4, 0.3]))
+array(0.16996714)
+
+However, deviating from this will result in recompilation and a warning message:
+
+>>> circuit(jnp.array([1.4, 1.4, 0.3, 0.1]))
+catalyst/compilation_pipelines.py:592:
+UserWarning: Provided arguments did not match declared signature, recompiling...
+Tracing occurring
+array(0.16996714)
+
+Try and compile the full workflow
+---------------------------------
+
+When porting your PennyLane code to work with Catalyst and :func:`@qjit <.qjit>`, the
+biggest performance advantage you will see is if you qjit
+your *entire* workflow, not just the QNodes. So think about putting
+everything inside your JIT-compiled function, including for loops
+(including optimization loops), gradient calls, etc.
+
+Consider the following PennyLane example, where we have a parametrized
+circuit, are measuring an expectation value, and are optimizing the result:
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=4)
+
+    @qml.qnode(dev)
+    def cost(weights, data):
+        qml.AngleEmbedding(data, wires=range(4))
+
+        for x in weights:
+            # each trainable layer
+            for i in range(4):
+                # for each wire
+                if x[i] > 0:
+                    qml.RX(x[i], wires=i)
+                elif x[i] < 0:
+                    qml.RY(x[i], wires=i)
+
+            for i in range(4):
+                qml.CNOT(wires=[i, (i + 1) % 4])
+
+        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))
+
+    weights = jnp.array(2 * np.random.random([5, 4]) - 1)
+    data = jnp.array(np.random.random([4]))
+
+    opt = jaxopt.GradientDescent(cost, stepsize=0.4)
+
+    params = weights
+    state = opt.init_state(params)
+
+    for i in range(200):
+        (params, _) = tuple(opt.update(params, state, data))
+
+Using PennyLane v0.32 on Google Colab with the Python 3 Google Compute Engine
+backend, this optimization takes 3min 28s ± 2.05s to complete.
+
+Let's switch over to Lightning, our high-performance statevector simulator,
```

Comment:
Would you like to link to the Lightning repo as you're adding a brief introduction here or this is too much? :D

Reply:
```suggestion
Let's switch over to [Lightning](https://docs.pennylane.ai/projects/lightning/en/stable/), our high-performance statevector simulator,
```